### PR TITLE
Support for both dynamic and static dispatch of events to transition state

### PR DESCRIPTION
**Pull request by jkonecki**
Dynamic dispatch uses `dynamic` keyword to utilize .NET dynamic method dispatch.
Static dispatch is based on explicit implementation of `IJournaledGrainStateTransition<TEvent>` interface.

### CODE REVIEW

File: src/Orleans/EventSourcing/JournaledGrainState.cs [hunk 1]
```diff
@@ -39,25 +40,45 @@ protected JournaledGrainState(Type type )
...
         public int Version { get; private set; }
 
-        public void AddEvent(StateEvent @event)
+        public void AddEvent<TEvent>(TEvent @event)
+            where TEvent: StateEvent
         {
             events.Add(@event);
-            ApplyEvent(@event);
+
+            DynamicStateTransition(@event);
```
Comment (sergeybykov) [inline src/Orleans/EventSourcing/JournaledGrainState.cs]: How do we choose which one to invoke?

**Comment by jkonecki** [comment]
Uncomment and recompile - I don't think we need both approaches in Orleans, we should choose one.
Unless for some reason we want flexibility here and in that case a simple interface and two providers will be needed alongside some configuration.

I've added static dispatcher as I remember someone was mentioning similar approach. I've been using dynamic dispatch in a couple of projects and it more than enough - the only question is how to handle missing `Apply` methods - NEventStore for example has a flag whether to throw an exception or just skip the event

**Comment by sergeybykov** [comment]
Thanks for the clarification. I wasn't sure if you meant to use both approaches, and pick one on the fly based on some criteria.

The dynamic option is elegant but type unsafe. Also, its performance cost under heavy use is unclear. The static option is safer but unnecessarily verbose.

**Comment by jkonecki** [comment]
We can use the reflection to check for existence of Apply methods and cache delegates.

Another way would be to generate MSIL to invoke Apply methods on the first execution, something dapper-dot-net (https://github.com/StackExchange/dapper-dot-net) is doing for deserialisation of db query results.

**Comment by ReubenBond** [comment]
Either way, the dynamic option will not be compile-time type safe. Emitting CIL is preferred over caching the results of reflection. Alternatively, Expression Trees will work unless the Apply() methods are private.

**Comment by jkonecki** [comment]
Regarding dynamic method calls performance:

http://theburningmonk.com/2010/09/performance-test-dynamic-method-invocation-in-csharp-4
http://byterot.blogspot.co.uk/2012/05/performance-comparison-of-code.html?m=1

TL;DR: dynamic is slightly slower than static but much faster than reflection

**Comment by jkonecki** [comment]
@reubenbond what do you mean by compile time safe? I'm thinking about scenarios where event stream contains events that are no longer needed by the state and therefore there are no Apply methods for them.

**Comment by ReubenBond** [comment]
@jkonecki "no longer needed" is indistinguishable from "not implemented", though.

I'm not saying it's bad - Scala typically uses partial functions rather than individual methods for handling messages, so the same deal there. Similarly for BEAM languages... so all of the popular frameworks wont tell you at compile time that you don't implement a handler anyway.

Actually, I like the `Apply(MyEventType event)` method - it's terse.

Oh, by the way, could we possibly split this into a separate NuGet package rather than including it in the core? I want the core to be slimmed down. Akka's event sourcing is a separate module.

**Comment by jkonecki** [comment]
I think a separate NuGet is a good idea.

**Comment by sergeybykov** [comment]
Maybe I'm missing something, but isn't the static option equally unsafe - depending on whether or not we have an `if` at https://github.com/dotnet/orleans/pull/780/files#diff-be5f9a34a52adb69d1df489479237a67R81, we'll either get a `NullReferenceException` or skip an "unimplemented" or "no longer needed event". Seems to me equivalent to the dynamic option with error handling. No?

**Comment by jkonecki** [comment]
I don't think you will get NRE. We will just skip the events that are not
handled (interface not implemented).

**Comment by sergeybykov** [comment]
Yes, because we have `if(transition != null)`. I meant that without the `if` we would get an NRE. Which seems morally equivalent to me to catching and ignoring a "bad cast" exception thrown by `me.Apply(this, @event);` in the dynamic case if `Apply` of that type is missing. From that perspective, I don't see that the static option is really safer.

**Comment by gabikliot** [comment]
Sergey, the static is safer in the following sense (assuming I actually understood what is going on here):
in the example here 
`public class StaticPersonState : JournaledGrainState<StaticPersonState>,  IJournaledGrainStateTransition<StaticPersonState, PersonRegistered> `
if this class does not have `public void Apply(StaticPersonState state, PersonRegistered registered) ` it will be compile time error, while with dynamic it will not be.

But on the other hand, withy static, if you forgot to add `IJournaledGrainStateTransition<StaticPersonState, PersonRegistered>` that it will be equivalent to dynamic. So bottom line: static can catch some cases, but not all.

File: src/TestGrains/EventSourcing/PersonState.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ MIT License
 
 namespace TestGrains
 {
-    public class PersonState : JournaledGrainState
+    public class PersonState : JournaledGrainState<PersonState>
```
Comment (gabikliot) [inline src/TestGrains/EventSourcing/PersonState.cs]: So just to make sure I understood correctly: this is an example of dynamic case? Can we then rename it to `DynamicPersonState`.

Comment (jkonecki) [inline src/TestGrains/EventSourcing/PersonState.cs]: Sure, I'll make another commit shortly

**Comment by jkonecki** [comment]
Just for a reference, here are two dispatch mechanisms (called Event Routers) from a NEventStore.Domain (formerly CommonDomain):

Convention (uses reflection)
https://github.com/NEventStore/NEventStore.Domain/blob/master/src/NEventStore.Domain/Core/ConventionEventRouter.cs

Explicit registration
https://github.com/NEventStore/NEventStore.Domain/blob/master/src/NEventStore.Domain/Core/RegistrationEventRouter.cs

I've been using convention based approach in multiple projects and never had any major problems. In an unlikely case of missing `Apply` method after adding new event all is needed is to add this method - new event has already been persisted in the event store so the state of the aggregate will 'correct itself' after missing `Apply` method is added.

**Comment by sergeybykov** [comment]
@gabikliot 
> if this class does not have  public void Apply(StaticPersonState state, PersonRegistered registered)  it will be compile time error, while with dynamic it will not be.

This doesn't really help because cast is done dynamically at run time. So it's still a convention - to remember what `IJournaledGrainStateTransition<T>` the class should implement, and this is not type safe.

The point I'm trying to make is that if the static approach isn't safer in the real scenarios of a missing handler for an event type, then the dynamic option seems better, and it's more elegant too. I'm not seeing any downside in going with just the dynamic option.

**Comment by jkonecki** [comment]
We could follow NEventStore.Domain approach an implement both as plugins
for users to choose from / extend.

**Comment by sergeybykov** [comment]
I think we could. Or we could start with a simple (convention) approach for now and grow it into the full blown plugin model later. Couldn't we?

**Comment by jkonecki** [comment]
That was my idea - go ahead with convention based approach and later
refactor is such a need arises. Refactoring will be pretty straightforward.

**Comment by sergeybykov** [comment]
I like that. Start simple, grow over time. Unless somebody sees a major issue here I'm not seeing.

**Comment by gabikliot** [comment]
I also prefer the dynamic approach, it is more terse.

**Comment by sergeybykov** [comment]
@jkonecki How about the following plan?
* Remove the static option with IJournaledGrainStateTransitions and all
* Merge the updated PR
* Rebase to master
* Move ES related types to a separate assembly (Orleans.EventSourcing.dll) and NuGet
* Prototype/define the storage provider story - whether ES can be used with the current provider API or needs a separate ES storage provider model

**Comment by jkonecki** [comment]
Sounds great.

I was thinking about the es provider. We probably was to be able to plug in
multiple event stores. Should we include a default one using Table Storage?

**Comment by sergeybykov** [comment]
> Should we include a default one using Table Storage?

That would be great.

**Comment by sergeybykov** [comment]
Oh, and feel free to get rid of StateEvent and switch to POCOs.

**Comment by jkonecki** [comment]
@yevhen has a nice library that implements Event Store over Table Storage.

https://github.com/yevhen/Streamstone

I think @gabikliot was looking at it. Any ideas if we can benefit from all
the great work @yevhen did?
I suppose adding a dependency on it to Orleans.EventSourcing may be too
much. But how about a separate NuGet: Orleans.EventSourcing.Streamstone ?

**Comment by sergeybykov** [comment]
I think we definitely want to leverage @yevhen's work. Doesn't Streamstone have its own NuGet package? Wouldn't we need a provider for Streamstone in Orleans.EventSourcing or in Orleans.EventSourcing.Providers?

**Comment by gabikliot** [comment]
I think the same `Orleans.EventSourcing` project can also have the `EventSourcing.Providers`, and also the provider to `Streamstone`. Starting with `Streamstone` is a good idea.

**Comment by sergeybykov** [comment]
I think it's cleaner to keep providers separate from ES core as providers are plugins.

**Comment by yevhen** [comment]
> Starting with Streamstone is a good idea

+1 :smile: 

Feel free to involve me :wink:

**Comment by jkonecki** [comment]
I just did ;-)

+1 for separate providers. Streamstone has a NuGet package. I can also
create a provider for GetEventStore as I'm familiar with it.
With two separate providers it will be easier to find common parts and
differences in order to design a reusable interface for event sourced
provider.

I'll make the changes in the branch in the next few days.

File: src/Orleans/EventSourcing/JournaledGrainState.cs [hunk 3]
```diff
@@ -28,36 +28,58 @@ MIT License
 
 namespace Orleans.EventSourcing
 {
-    public abstract class JournaledGrainState : GrainState
+    public abstract class JournaledGrainState<TGrainState> : GrainState
+        where TGrainState : JournaledGrainState<TGrainState>
     {
         private List<StateEvent> events = new List<StateEvent>();
-        protected JournaledGrainState(Type type )
-            : base(type.FullName)
```
Comment (jason-bragg) [inline src/Orleans/EventSourcing/JournaledGrainState.cs]: If this is a 'GrainState' shouldn't it support the same construction pattern as grain state?  I.E. JournaledGrainState() and JournaledGrainState(string name)

File: src/Orleans/EventSourcing/JournaledGrainState.cs [hunk 4]
```diff
@@ -28,36 +28,58 @@ MIT License
...
+
+        protected JournaledGrainState()
+            : base(typeof(TGrainState).FullName)
```
Comment (jason-bragg) [inline src/Orleans/EventSourcing/JournaledGrainState.cs]: If this is a 'GrainState' shouldn't it support the same construction pattern as grain state? I.E. JournaledGrainState() and JournaledGrainState(string name)

File: src/Orleans/EventSourcing/JournaledGrainState.cs [hunk 5]
```diff
@@ -28,36 +28,58 @@ MIT License
...
+
+            foreach (StateEvent @event in Events)
+                me.Apply(@event);
+        }
+
+        private void DynamicStateTransition<TEvent>(TEvent @event)
+            where TEvent : StateEvent
+        {
+            dynamic me = this;
+            me.Apply(this, @event);
```
Comment (jason-bragg) [inline src/Orleans/EventSourcing/JournaledGrainState.cs]: Not used dynamic enough to speak from experience, but my understanding is that dynamic per event would not be cheep (as compared to a switch, or other methods).

If a new TEvent type was added without a corresponding handler, the code would add the bad event to the event list, but fail to adjust the state, as well as causing the AddEvent to throw.  If the events were eventually persisted somewhere, this could cause some icky bugs.

Suggest considering some sort of handler registration?  Shouldn't add much to developer overhead, while allowing for better error handling as well as faster event handling.

File: src/Orleans/EventSourcing/JournaledGrainState.cs [hunk 6]
```diff
@@ -28,36 +28,58 @@ MIT License
...
+        }
+
+        private void StaticStateTransition<TEvent>(TEvent @event)
+            where TEvent : StateEvent
+        {
+            var transition = this as IJournaledGrainStateTransition<TGrainState, TEvent>;
+
```
Comment (jason-bragg) [inline src/Orleans/EventSourcing/JournaledGrainState.cs]: So for a grainstate to support 20 different types of TEvent, it would have to inherit from 20 different specializations of the IJournaledGrainStateTransition interface?  Wouldn't some simple type to handler map, with some basic handler registration logic be better?  It could be used with composition or inheritance.

Comment (gabikliot) [inline src/Orleans/EventSourcing/JournaledGrainState.cs]: Jason, this is what we called a static approach, which we agreed not to do. The consensus as of now is to go with dynamic approach (see examples for both approaches in this PR and in the tests). Your other comment about potentially fragility of the dynamic approach is a very valid one, which partly bothers us all. What is the concrete suggestion to improve it? How do you do "handler registration" to a (lets not forget **virtual**) grain ?

Comment (jkonecki) [inline src/Orleans/EventSourcing/JournaledGrainState.cs]: Also please take a note that it's the state that's event sourced not the
actual grain.
